sdk/core/azure-core-http-jdk-httpclient/src/test/java/com/azure/core/http/jdk/httpclient/JdkAsyncHttpClientBuilderTests.java

```diff
@@ -231,4 +242,104 @@ public void buildWithNonProxyConfigurationProxy() {
+
+        validateRestrictedHeaders(jdkAsyncHttpClientBuilder, expectedRestrictedHeaders, 3);
+    }
+
+    @Test
+    void testAllowedHeadersFromConfiguration() {
+        Configuration configuration = Configuration.getGlobalConfiguration();
+        configuration.put("jdk.httpclient.allowRestrictedHeaders", "content-length, upgrade");
+
+        JdkAsyncHttpClientBuilder jdkAsyncHttpClientBuilder = spy(
+            new JdkAsyncHttpClientBuilder().configuration(configuration));
+
+        Properties properties = new Properties();
+        when(jdkAsyncHttpClientBuilder.getNetworkProperties()).thenReturn(properties);
+
+        Set<String> expectedRestrictedHeaders = new TreeSet<>(String.CASE_INSENSITIVE_ORDER);
+        expectedRestrictedHeaders.addAll(JdkAsyncHttpClientBuilder.DEFAULT_RESTRICTED_HEADERS);
+        expectedRestrictedHeaders.removeAll(Arrays.asList("content-length", "upgrade"));
+
+        validateRestrictedHeaders(jdkAsyncHttpClientBuilder, expectedRestrictedHeaders, 3);
+    }
+
+    @Test
+    void testAllowedHeadersFromBoth() {
+        Configuration configuration = Configuration.getGlobalConfiguration();
+        configuration.put("jdk.httpclient.allowRestrictedHeaders", "content-length, upgrade");
+
+        JdkAsyncHttpClientBuilder jdkAsyncHttpClientBuilder = spy(
+            new JdkAsyncHttpClientBuilder().configuration(configuration));
+
+        Properties properties = new Properties();
+        properties.put("jdk.httpclient.allowRestrictedHeaders", "host, connection, upgrade");
+        when(jdkAsyncHttpClientBuilder.getNetworkProperties()).thenReturn(properties);
+
+        Set<String> expectedRestrictedHeaders = new TreeSet<>(String.CASE_INSENSITIVE_ORDER);
+        expectedRestrictedHeaders.addAll(JdkAsyncHttpClientBuilder.DEFAULT_RESTRICTED_HEADERS);
+        expectedRestrictedHeaders.removeAll(Arrays.asList("content-length", "host", "connection", "upgrade"));
+
+        validateRestrictedHeaders(jdkAsyncHttpClientBuilder, expectedRestrictedHeaders, 1);
+    }
+
+    @Test
+    void testAllowedHeadersFromSystemProperties() {
+        System.setProperty("jdk.httpclient.allowRestrictedHeaders", "content-length, upgrade");
+
+        JdkAsyncHttpClientBuilder jdkAsyncHttpClientBuilder = spy(new JdkAsyncHttpClientBuilder());
+        Properties properties = new Properties();
+        when(jdkAsyncHttpClientBuilder.getNetworkProperties()).thenReturn(properties);
+
+        Set<String> expectedRestrictedHeaders = new TreeSet<>(String.CASE_INSENSITIVE_ORDER);
+        expectedRestrictedHeaders.addAll(JdkAsyncHttpClientBuilder.DEFAULT_RESTRICTED_HEADERS);
+        expectedRestrictedHeaders.removeAll(Arrays.asList("content-length", "upgrade"));
+
+        validateRestrictedHeaders(jdkAsyncHttpClientBuilder, expectedRestrictedHeaders, 3);
+    }
+
+    @Test
+    void testCaseInsensitivity() {
+        System.setProperty("jdk.httpclient.allowRestrictedHeaders", "content-LENGTH");
```

Comment:
If you set this system property, after the test it should be returned to its original state.